lib/rpc/UniJsonRpcProvider.ts

```diff
@@ -0,0 +1,316 @@
+import { SingleJsonRpcProvider } from './SingleJsonRpcProvider'
+import { StaticJsonRpcProvider, TransactionRequest } from '@ethersproject/providers'
+import { isEmpty } from 'lodash'
+import { ChainId } from '@uniswap/sdk-core'
+import {
+  BlockTag,
+  BlockWithTransactions,
+  Filter,
+  Log,
+  TransactionReceipt,
+  TransactionResponse,
+} from '@ethersproject/abstract-provider'
+import { LRUCache } from 'lru-cache'
+import { BigNumber, BigNumberish } from '@ethersproject/bignumber'
+import { Deferrable } from '@ethersproject/properties'
+import Logger from 'bunyan'
+
+export class UniJsonRpcProvider extends StaticJsonRpcProvider {
+  private readonly chainId: ChainId = ChainId.MAINNET
+
+  private readonly providers: SingleJsonRpcProvider[] = []
+
+  // Used to remember the user-specified precedence or provider URLs. 0 means highest
+  private readonly urlPrecedence: Record<string, number> = {}
+
+  // If provided, we will use this weight to decide the probability of choosing
+  // one of the healthy providers.
+  private readonly urlWeight: Record<string, number> = {}
+
+  private lastUsedProvider: SingleJsonRpcProvider | null = null
+
+  private totallyDisableFallback: boolean = false
+
+  private sessionCache: LRUCache<string, SingleJsonRpcProvider> = new LRUCache({ max: 1000 })
+
+  // If true, it's allowed to use a different provider if the preferred provider isn't healthy.
+  private allowProviderAutoSwitch: boolean = true
+
+  private readonly log: Logger
+
+  constructor(
+    chainId: ChainId,
+    singleRpcProviders: SingleJsonRpcProvider[],
+    log: Logger,
+    ranking?: number[],
+    weights?: number[],
+    allowProviderAutoSwitch?: boolean
+  ) {
+    // Dummy super constructor call is needed.
+    super('dummy_url', { chainId, name: 'dummy_network' })
+    this.connection.url
+    this.log = log
+
+    if (isEmpty(singleRpcProviders)) {
+      throw new Error('Empty singlePrcProviders')
+    }
+
+    if (ranking !== undefined && weights !== undefined && weights && ranking.length != weights.length) {
```

Comment:
nit: `ranking?.length !== weights?.length`

It will also cover the 3 permutations of ranking is undefined and weights is undefined, which are all not going to throw error.